[Siddanth Umralkar] iP

wowsiddanth opened the pull request:
# Petal 🌸
> *In joy or sadness, flowers are our constant friends.*
> **Okakura Kakuzo**

It's okay. Petal is here. Drop your worries and sink into the petals of relief; Petal is here for you.

Petal has the following abetments: 
* Text-based 
* Simple commands
* Task tracking

## Installation

A logical thing you might do is wonder—*how do I install Petal*?

There are two methods:
      
### 1. Run it in an IDE
1. Go to this [link](https://github.com/wowsiddanth/ip).
2. Download the folder.
3. Open it in IntelliJ or your favorite IDE.
4. Then, under the Petal class. add the following main method:
```java
public static void main(String[] args) {
        Petal p = new Petal();
        p.run(); 
 }
```
### 2. Run the .jar file
1. Go to this [link](https://github.com/wowsiddanth/ip).
2. Under the __Releases__ tab, open __A-Jar__ and download the .jar file.
3. Download the folder.
4. Open command prompt from that folder, or your favorite IDE.
5. Then, type this into the command prompt: `java -jar iP.jar`

## Features
Petal can:
- [X] Keep track of tasks
- [X] Keep track of the dates of tasks, including deadlines and events.
- [ ] Set reminders (coming soon)

Commit:
Add Gradle support

wowsiddanth committed:
- First level, first iteration. Added every functionality required.

wowsiddanth committed:
Level 2 - Iteration 1. Implemented required features

wowsiddanth committed:
Task 3 - Iteration 1. Worked on all requirements. B

wowsiddanth committed:
Revert "Task 3 - Iteration 1. Worked on all requirements. B"

This reverts commit 36d22d189c25c068b1564048721f13939f9d9401.

wowsiddanth committed:
Level 3 - Iteration 1 Worked on all requirements.

wowsiddanth committed:
Task 3 - Iteration 1 Worked on all requirements

wowsiddanth committed:
Level 3 - Iteration 1 Worked on all requirements

wowsiddanth committed:
Level 3 - Iteration 2 Polished certain features. Changed switch statement to if/else and seperate print statement etcs. into their own methods.

wowsiddanth committed:
Level 4 - Iteration 1 Basic features added. No functionality yet.

wowsiddanth committed:
Level 4 - Iteration 2 Improved functionality

wowsiddanth committed:
Level 4 - Iteration 3 Added some error management and polished features

wowsiddanth committed:
Level 4 - Iteration 4 Updated the Expected.txt file. Conducted A-TextUiTesting

wowsiddanth committed:
Level 4 - Iteration 5 Update test files. Update name.

wowsiddanth committed:
Level 5 - Iteration 1 Add basic implementation. Add exceptions

wowsiddanth committed:
Level 5 - Iteration 2 Change try/catch block. Minor improvements

wowsiddanth committed:
Level 6 - Iteration 1 Add delete. Minor improvements

wowsiddanth committed:
Level 6 - Iteration 2 Remove some statements from Try/Catch

wowsiddanth committed:
Level 6 - Iteration 3 Change PetalException message

wowsiddanth committed:
Level 6 - Iteration 4 Update major changes

wowsiddanth committed:
Level 6 - Iteration 4 Update major changes

wowsiddanth committed:
Level 6 - Iteration 5 Add JavaDocs, New method for processing request

wowsiddanth committed:
Level 6 - Iteration 6
Made changes to certain how handleTask works. Add trimming to messages to allow messages with whitespaces to still work.

wowsiddanth committed:
Level 6 - Iteration 7
Made minor changes to the documentation.
Changed the exception thrown by the deleteTask() method.

wowsiddanth committed:
Level 6 - Iteration 8
Update logic behind message handling.
Move scanner to within the Petal class
Wrap Java exceptions with own exception class for Java exceptions

wowsiddanth committed:
Level 6 - Iteration 9
Update exception to act like wrapper exceptions as well
Change code to wrap some exception as PetalException and its sub exceptions
Add test cases to test
Update EXPECTED.txt to check as well

wowsiddanth committed:
Level 6 - Iteration 10
Add capital words to be used as well
Add method to capitalize first letter for output for nicer output
Remove string/cause fields for custom exceptions
Add more test cases to test capitalized input, and wrong format

wowsiddanth committed:
Level 6 - Iteration 11
Change description of Petal class
Add A-Collections, A-Inheritance, A-Exceptions, A-Classes tags as these were not added before

wowsiddanth committed:
Level 6 - Iteration 12
Add test cases and change Expected accordingly
Change name of Indentation to LINE
Rename start() to run()
Add function printMessage() to handle output
Change name of formatMessage() to handleInput()

wowsiddanth committed:
Level 6 - Iteration 12
Minor bug fixes, where first letter of time was not capitalized

wowsiddanth committed:
Level 6 - FINAL
Minor changes to format of done/task add messages
Change descriptions of classes to be more descriptive
Add more test cases to input for thorough testing
Changed expected accordingly

wowsiddanth committed:
Level 6 - FINAL
Minor changes

wowsiddanth committed:
Level 6 - FINAL
Update Expected.txt

wowsiddanth committed:
Add basic saving features

The bot does not have any way to save the given tasks. This results in the user having
to repeatedly enter his tasks.

Add saving mechanisms. Create the relevant folder and file (createDirectory()). Write the
tasks into the .txt file (retrieveTasks()) and read from it when the bot start (if possible).

It is being done this way as having multiple methods handle the writing/reading aspect
increasing debugging ability and readability.

wowsiddanth committed:
Format and abstract response, fix bugs

Some responses are quite long and cluttered up the Petal class, reducing readability.
The strings returned by some methods (Task toString()) and printList() are not in the
format required for saving. There are bugs in the displayed list after accessing saved tasks
because of the wrong format being used to save the tasks.

Add abstract methods strForSaving() that returns a string representation for the task in the
proper format (e.g. T|X|go for a run). Add formatForSaving() that prints using this method.

Used enums as it always for constants and used abstract methods to take advantage
of polymorphism.

wowsiddanth committed:
Add time feature and add test cases

The code can only handle time in a string form, and is unable to do anything more with it.

Store time as a LocalDate and LocalTime object. Add additional test cases to test for bugs
in time input.

Storing time as a LocalDate and LocalTime object gives much more flexibility. allowing
Petal to display time in a specific format, and in the future, track events based on the data

wowsiddanth committed:
Retrieve tasks on given date (deadline)

Petal can only associate a deadline with a given task, but cannot return an output
with all the deadline on a given day.

Add HashMap. Use static method that takes in a date and returns string
representation of value stored at date.

The HashMap allows for the use of key-value pairs that will store the corresponding dead-
lines at that date. Public static method will allow the method to be called from the Petal
class and return the string representation.

wowsiddanth committed:
Retrieve events and deadlines on given date

The code can only retrieve deadlines on a specified date, but not events.

Add similar methods to event class to parse given input into date, start time, and end time.
Use hashmap to keep track of events on that date.

Similar reasons to the deadline commit. Refers to the prefer commit for more information.

wowsiddanth committed:
For level 8

wowsiddanth committed:
Add more OOP

The bot now does everything via the Petal class and few other classes. Many process can
be abstracted under their own classes.

Create 4 new classes: Parser, Storage, TaskList, and Ui. Separate methods and fields into
respective classes.

Separating the bots into these classes allows me to model the bot as the interactions
between these components.

wowsiddanth committed:
Organize into packages

Program now is using the default package.

Create different packages (petal, petal.exception, petal.task, petal.components).
Separate classes into respective packages.

Packages will lead to easier management.

wowsiddanth committed:
Update test cases and change runtest.bat

Outdated test cases and runtest doesn't work because of packages

Update test cases. Add paths of packages.

wowsiddanth committed:
Abstract more OOP

Code still had elements that could be abstracted out, especially commands.

Create more class corresponding to their specific function, like ByeCommand etc. Place
these classes in the Command package.

It allows for:
 * the code specific to this command to be abstracted out to its own class and
   added to the command package
 * Takes advantage of runtime polymorphism as the execute() is
   overridden and has different effects at runtime.

wowsiddanth committed:
Fix bugs and move scanner

After abstraction, the bye command no longer works.

Add public method in Petal to terminate bot. Move scanner to Ui class as part of
OOP.

The public method allows the Ui class to toggle the boolean Bye, turning it off and moving
the scanner to Ui allows for better abstraction.

wowsiddanth committed:
Add junit tests and clean up imports

No junit tests and some unused imports statements.

Create junit tests and corresponding stubs for unit testing. Delete import statements.

Stubs allow for unit testing by isolating classes.

wowsiddanth committed:
Change line characters

Flower characters cannot be displayed properly on terminal

Change characters from 'ꕤ' to '='

wowsiddanth committed:
Add javadoc, fix bug, and abstract calendar class

Some methods do not have Javadocs. Bug where delete task did not disappear from
the date. The way tasks are stored with dates is inefficient.

Add documentation for the remaining methods. Fix the bug related to the task.
Add calendar class and perform date-related operation using it..

Adding a calendar class allows for more abstraction.

wowsiddanth committed:
Add javadoc

wowsiddanth committed:
Follow the coding standard

Code partially follows the coding standards.

Adjust the code to follow the coding standard.

wowsiddanth committed:
Follow coding standard

Code partially follows coding standard

Adjust code to follow coding stamdard

wowsiddanth committed:
Add find

Cannot find for task based on keyword

Create method to find for task based on keyword

wowsiddanth committed:
Add find

Cannot find for task based on keyword

Create method to find for task based on keyword

wowsiddanth committed:
Change access modifier of stub bool

Wrong access modifier for isDone in TaskStub class

Change access modifier and add new field called isDone within
the TaskStub class itself

The class now has its own isDone field and no longer needs
to access that of its super class

wowsiddanth committed:
Improve code naming

Code still has some names/style that can be
improved.

Change the code to adhere better.

wowsiddanth committed:
Merge grandle support branch

wowsiddanth committed:
Add gradle support

Code does not allow for gradle support.

Modifiy build.gradle to allow Petal as main class. Set up
gradle for project.

wowsiddanth committed:
Adjust response to timeable exception

wowsiddanth committed:
Merge branch with grandle support

wowsiddanth committed:
Utilize checkstyle

Code was still incompatible to style guide in some areas

Use checkstyle to check and correct mistakes

vimuthm commented on src/main/java/petal/Petal.java:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    /**
+     * Flips the boolean isBye and terminates the process
```
It might be better to comment, "Sets the boolean isBye to false..." to remove possible confusion arising from using the word "flip"

vimuthm commented on src/main/java/petal/components/Calendar.java:
```diff
@@ -0,0 +1,77 @@
+        if (current.isPresent()) {
+            current.get().add(timeable);
+        } else {
+            ArrayList<Timeable> firstList = new ArrayList<>();
```
Coding Standard mentions:

> Plural form should be used on names representing a collection of objects.

`firstList` might not strictly adhere to this but I'm not sure if it's an outright violation. What do you think?

vimuthm commented on src/main/java/petal/components/Storage.java:
```diff
@@ -0,0 +1,122 @@
+        try {
+            File tasks = new File(filePath);
+            Scanner scanner = new Scanner(tasks);
+            ArrayList<Task> toBeAdded = new ArrayList<>();
```
Same issue as previous collections variable naming

vimuthm commented on src/main/java/petal/task/ToDo.java:
```diff
@@ -0,0 +1,38 @@
+
+/**
+ * The To.Do class that is a subclass of Task
+ * (Used a . in between to prevent IntelliJ from highlighting it)
```
Nice :) I faced the same issue!

vimuthm reviewed (commented):
Overall really good use of OOP and nice job on the Calendar feature!

One thing I noticed with the design implementation was a persisting check on todo/deadline/event. Within `Parser.java`, all 3 cases are grouped to return a `TaskCommand` which then calls upon `handleTask` in `TaskList.java` where you parse it to different cases again.

Personally, I feel maybe having a subclass of `TaskCommand` to handle these separately might be better. After which, handling the command input parsing seems more aligned to the duties of the corresponding command class and not within `TaskList` (if command format changes for example, current implementation means you need to modify the `TaskList` methods and not the `TaskCommand` class)

Any thoughts on this?

wowsiddanth committed:
Add GUI

Code was a CLI application

Add GUI, using JavaFX and FXML.

With a GUI, it becomes more user-friendly and is more intuitive for
less technically-inclined users.

wowsiddanth committed:
Add GUI

Code was a CLI application

Add GUI, using JavaFX and FXML.

With a GUI, it becomes more user-friendly and is more intuitive for
less technically-inclined users.

wowsiddanth committed:
Make minor changes to user image

wowsiddanth committed:
Update README.md

wowsiddanth committed:
Update README.md

wowsiddanth committed:
Update main method class

Main method is in the Launcher class.

Change main method to be in Petal class.

This is
* more logical
* application can be run via Gradle.

wowsiddanth committed:
Merge remote branch with updated readme

wowsiddanth committed:
Place Ui between Petal and MainWindow classes

Main Window (part of GUI) communicates directly with Petal.

Place Ui in between petal and MainWindow and create double association between petal
and Ui, and Ui and MainWindow class.

This is because the MainWindow class should only communicate to Petal via the
UI class, that is designed to handle user input/petal output. Placing
the Ui class between these two classes allows the code to follow Ui's
purpose.

wowsiddanth committed:
Change JUnit test cases and fix style

The JUnit test cases are outdated.

Update JUnit test cases to comply with
GUI.

This allows for the continued testing of the
Petal and other classes.

wowsiddanth committed:
Add assertions

The code does not have any assertions, and this may cause issues later now as
the states of certain attributes and objects is assumed (not verified) which may cause
unwanted side effects.

Add assertions liberally throughout the code (especially to assert if a attribute is not null,
for example in the case of 'assert mainWindow != null' in the Ui class.

It is being done this way in accordance with the documentation, which states that asserts
should test for certain states in the program.

wowsiddanth committed:
Merge pull request #1 from wowsiddanth/branch-A-Assertions

Add assertions

wowsiddanth committed:
Refactor code and check code quality

Code is bloated and less readable in some areas. Although it works as intended,
the code can be modified to improve overall code quality.

Refactor code by following given conventions. Extract out methods and attributes and
remove logic that is not required in the code.

It is being done that way to improve readability. Here are some of the changes done:
* Extract method from hasUsedPetalBefore() called readTasks(...)
* Extract out components array from readTasks()
* Change variable isLocationPresent to isFolderAndFilePresent

wowsiddanth committed:
Add streams

The code does not make use of streams and some methods are very long and can be
simplified.

Add streams to replace certain method bodies like the findTasksWithKeyword().

This improves readability and it shortens the method body and the same
functionality can be summarized using streams.

wowsiddanth committed:
Merge pull request from CodeQuality

Refactor code and check code quality

wowsiddanth committed:
Merge branch 'master' into branch-A-Streams

wowsiddanth committed:
Merge pull request #3 from wowsiddanth/branch-A-Streams

Add streams

wowsiddanth committed:
Add archive function

Code does not allow you to archive current tasks and subsequently, cannot start
again with a clean slate without losing all previous progress.

Add archive functionality by using the DossiersCommand, ArchiveCommand class. Add
relevant methods to TaskList and storage class.

It is being done this way in accordance with the existing OOP format, and it allows
for more flexibility to change in the future.

The archived tasks cannot be merged with the current tasks, for that functionality
the user still has to open the text file with the archived tasks and manually shift them to the
text file of current tasks.

wowsiddanth committed:
Add extensive JUnit testing

Current testing only meets the requirements but does not allow me to carry out
extensive testing on different components of Petal for regression and unit testing
purposes.

Add more JUnit test cases. Add more test classes that test the main classes, such as
CalendarTest, TaskListTest, and TaskTest.

It was done this way to ensure that JUnit works and every class can be tested in
isolation (with stub classes).

wowsiddanth committed:
Add small style changes

wowsiddanth committed:
Add CI

Code does not have GitHub's Continuous Integration for workflows.

Add gradle.yml to enable CI.

Done in accordance to requirements.

wowsiddanth committed:
Merge branch 'master'

wowsiddanth committed:
Set theme jekyll-theme-dinky

wowsiddanth committed:
Set theme jekyll-theme-cayman

wowsiddanth committed:
Edit README.md

wowsiddanth committed:
Merge branch 'master' of https://github.com/wowsiddanth/ip

wowsiddanth committed:
Edit config.yml to add baseURL

wowsiddanth committed:
Remove config.yml and images

wowsiddanth committed:
Add .nojekyll file

wowsiddanth committed:
Add minor change to README.md

wowsiddanth committed:
Modify README.md

wowsiddanth committed:
Update README.md

wowsiddanth committed:
Update README.md

wowsiddanth committed:
Rename screenshot to Ui.png

wowsiddanth committed:
Merge branch 'master' of https://github.com/wowsiddanth/ip

wowsiddanth committed:
Update README.md

wowsiddanth committed:
Update README.md

wowsiddanth committed:
Update README.md

wowsiddanth committed:
Replace Ui.png

wowsiddanth committed:
Update README.md

wowsiddanth committed:
Add PictureEditor class, and remove unnecessary code

Code has potential to have another class, PictureEditor class abstracted out, that
deals with the editing of the pictures to be shown. Additional, there is some code
that is redundant and that can be removed.

Extract PictureEditor class, and clean up code.

This is done to improve OOP and increase readability of the code overall.

wowsiddanth committed:
Fix date bug and minor style changes

Entering date without anything else (e.g. date) results in an InvocationException, that
does not crash the program but just leaves a blank in the GUI.

Fix the date bug by catching ArrayIndexOutOfBoundsException.

wowsiddanth committed:
Update README.md

wowsiddanth committed:
Update Ui.png

wowsiddanth committed:
Rename Ui.png

wowsiddanth committed:
Rename Ui.png

wowsiddanth committed:
Rename Ui.png

wowsiddanth committed:
Update README.md
